storage/integration_test.go

```diff
@@ -2573,259 +2576,348 @@ func TestIntegration_BucketIAM(t *testing.T) {
+				_, err = bucket.DefaultObjectACL().List(ctx)
+				checkforErrors("default object acl list", err)
+				checkforErrors("default object acl delete", bucket.DefaultObjectACL().Delete(ctx, entity))
+
+				// Copy
+				_, err = bucket.Object("copy").CopierFrom(bucket.Object(objectName)).Run(ctx)
+				checkforErrors("copy", err)
+				// Delete "copy" object, if created
+				if err == nil {
+					t.Cleanup(func() {
+						h.mustDeleteObject(bucket.Object("copy"))
+					})
+				}
 
+				// Compose
+				_, err = bucket.Object("compose").ComposerFrom(bucket.Object(objectName), bucket.Object("copy")).Run(ctx)
+				checkforErrors("compose", err)
+				// Delete "compose" object, if created
+				if err == nil {
+					t.Cleanup(func() {
+						h.mustDeleteObject(bucket.Object("compose"))
+					})
+				}
+
+				// Delete object
+				if err = bucket.Object(objectName).Delete(ctx); err != nil {
+					// We still want to delete object if the test errors
+					h.mustDeleteObject(requesterPaysBucket.Object(objectName))
+				}
+				checkforErrors("delete object", err)
+			})
+		}
+	})
+}
+
+// This test needs a second project and user to test all possibilities. Since we
+// need these things for Firestore already, we use them here.
+//
+// There are up to three entities involved in a requester-pays call:
+//  1. The user making the request. Here, we use the account used for the
+//     Firestore tests. The following must hold for this test to work:
+//     - this user must NOT have resourcemanager.projects.createBillingAssignment
+//     on the project that owns the bucket (2).
+//     - this user must have serviceusage.services.use permission on the Firestore
+//     project (3b).
+//     - this user must NOT have that serviceusage.services.use permission on
+//     the project that owns the bucket (3a).
+//  2. The project that owns the requester-pays bucket. Here, that
+//     is the test project ID (see testutil.ProjID).
+//  3. The project provided as the userProject parameter of the request;
+//     the project to be billed. This test uses:
+//     a. The project that owns the requester-pays bucket (same as (2))
+//     b. Another project (the Firestore project).
+func TestIntegration_RequesterPaysNonOwner(t *testing.T) {
 	ctx := context.Background()
 
-	// Start client with mainUserEmail creds
-	mainUserClient := testConfig(ctx, t)
-	defer mainUserClient.Close()
+	// Main project: the project that owns the requester-pays bucket.
+	mainProject := testutil.ProjID()
 
-	// Get user emails from credentials
-	jwt, err := testutil.JWTConfig()
-	if err != nil {
-		t.Fatalf("testutil.JWTConfig: %v", err)
+	// Secondary project: a project that does not own the bucket.
+	// The "main" user does not have permission on this.
+	// This project should have billing enabled.
+	secondaryProject := os.Getenv(envFirestoreProjID)
+	if secondaryProject == "" {
+		t.Fatalf("need a second project (env var %s)", envFirestoreProjID)
```

Comment:
The check for `-short` needs to come before this, otherwise a unit test in short mode will fail. Since this is usually in `testConfig` you can add an explicit check instead to avoid having to move all this into the transport specific tests.